geo/src/algorithm/sweep/line_or_point.rs

```diff
@@ -0,0 +1,259 @@
+use std::cmp::Ordering;
+
+use super::SweepPoint;
+use crate::{
+    line_intersection::line_intersection, Coordinate, GeoFloat, GeoNum, HasKernel, Kernel, Line,
+    LineIntersection, Orientation,
+};
+
+/// Either a line segment or a point.
+///
+/// The coordinates are ordered (see [`SweepPoint`]) and a line
+/// segment must have distinct points (use the `Point` variant if the
+/// coordinates are the equal).
+#[derive(Clone, Copy)]
+pub struct LineOrPoint<T: GeoNum> {
+    left: SweepPoint<T>,
+    right: SweepPoint<T>,
+}
+
+impl<T: GeoNum> std::fmt::Debug for LineOrPoint<T> {
+    fn fmt(&self, f: &mut std::fmt::Formatter<'_>) -> std::fmt::Result {
+        f.debug_tuple(if self.is_line() { "LPt" } else { "Pt" })
+            .field(&self.left.x_y())
+            .field(&self.right.x_y())
+            .finish()
+    }
+}
+
+impl<T: GeoNum> From<SweepPoint<T>> for LineOrPoint<T> {
+    fn from(pt: SweepPoint<T>) -> Self {
+        Self {
+            left: pt,
+            right: pt,
+        }
+    }
+}
+
+impl<T: GeoNum> From<(SweepPoint<T>, SweepPoint<T>)> for LineOrPoint<T> {
+    fn from(pt: (SweepPoint<T>, SweepPoint<T>)) -> Self {
+        let (start, end) = pt;
+        match start.cmp(&end) {
+            Ordering::Less => Self {
+                left: start,
+                right: end,
+            },
+            _ => Self {
+                left: end,
+                right: start,
+            },
+        }
+    }
+}
+
+/// Convert from a [`Line`] ensuring end point ordering.
+impl<T: GeoNum> From<Line<T>> for LineOrPoint<T> {
+    fn from(l: Line<T>) -> Self {
+        let start: SweepPoint<T> = l.start.into();
+        let end = l.end.into();
+        (start, end).into()
+    }
+}
+
+/// Convert from a [`Coordinate`]
+impl<T: GeoNum> From<Coordinate<T>> for LineOrPoint<T> {
+    fn from(c: Coordinate<T>) -> Self {
+        Self {
+            left: c.into(),
+            right: c.into(),
+        }
+    }
+}
+
+impl<T: GeoNum> LineOrPoint<T> {
+    /// Checks if the variant is a line.
+    #[inline]
+    pub fn is_line(&self) -> bool {
+        self.left != self.right
+    }
+
+    /// Return a [`Line`] if it is one, otherwise `None`.
+    #[inline]
+    pub fn line(&self) -> Line<T> {
+        Line::new(*self.left, *self.right)
+    }
+
+    #[inline]
+    pub fn left(&self) -> SweepPoint<T> {
+        self.left
+    }
+
+    #[inline]
+    pub fn right(&self) -> SweepPoint<T> {
+        self.right
+    }
+
+    #[cfg(test)]
+    pub fn coords_equal(&self, other: &LineOrPoint<T>) -> bool {
+        self.is_line() == other.is_line() && self.end_points() == other.end_points()
```

Comment:
Sorry, `end_points` is as used in math (so both vertices of the line are "end points").  It's only used in tests, because the `PartialEq` has to agree with `PartialOrd` and accepts more segments as being "equal" (anything that overlaps)

Reply:
oh right - end_point**s** I read it as `end_point`! My mistake.